### PR TITLE
Optimize Span.GetPinnableReference

### PR DESCRIPTION
**Pull request by jkotas**
Related to discussion in https://github.com/dotnet/corefx/pull/32669

`Span.GetPinnableReference` compiles into one conditional branch and one unconditional branch on the hot path. This change tweaks the implementation to make it compile to just one well-predicted conditional branch on the hot path.

Test (extracted from https://github.com/eerhardt/machinelearning/commit/56b849c6d54bfd9e73d8bec8c119660d3e5b6dc1#diff-0f95f507da073f50b5077ef70802a793R1191 - thanks @eerhardt!) :
```
    [MethodImpl(MethodImplOptions.NoInlining)]
    static float DotU(float*p, float* q, int count) => 0;

    [MethodImpl(MethodImplOptions.NoInlining)]
    public static float DotProductDense(ReadOnlySpan<float> a, ReadOnlySpan<float> b, int count)
    {
        fixed (float* pa = a)
        fixed (float* pb = b)
            return DotU(pa, pb, count);
    }

    static void DoStuff()
    {
        float[] a = new float[100]; 
        float[] b = new float[100];
        for (int i = 0; i < 1000000000; i++) DotProductDense(a, b, 100);
    }
```

Before this change: 4800ms
After this change: 4110ms
(for reference) Pinning using MemoryMarshal.GetReference: 3930ms

Disassembly for `DotProductDense` method before:
```
sub     rsp,38h
vzeroupper
xor     eax,eax
mov     qword ptr [rsp+30h],rax
mov     qword ptr [rsp+28h],rax
cmp     dword ptr [rcx+8],0
jne     test!Test.DotProductDense(System.ReadOnlySpan`1<Single>, System.ReadOnlySpan`1<Single>, Int32)+0x1cd
xor     eax,eax
jmp     test!Test.DotProductDense(System.ReadOnlySpan`1<Single>, System.ReadOnlySpan`1<Single>, Int32)+0x1d0
mov     rax,qword ptr [rcx]
mov     qword ptr [rsp+30h],rax
cmp     dword ptr [rdx+8],0
jne     test!Test.DotProductDense(System.ReadOnlySpan`1<Single>, System.ReadOnlySpan`1<Single>, Int32)+0x1df
xor     ecx,ecx
jmp     test!Test.DotProductDense(System.ReadOnlySpan`1<Single>, System.ReadOnlySpan`1<Single>, Int32)+0x1e2
mov     rcx,qword ptr [rdx]
mov     qword ptr [rsp+28h],rcx
mov     rdx,rcx
mov     rcx,qword ptr [rsp+30h]
call    Test.DotU(Single*, Single*, Int32) (00007fff`28ac1218)
nop
add     rsp,38h
ret
```

Disassembly for `DotProductDense` method after:
```
sub     rsp,38h
vzeroupper
xor     eax,eax
mov     qword ptr [rsp+30h],rax
mov     qword ptr [rsp+28h],rax
xor     eax,eax
cmp     dword ptr [rcx+8],0
je      test!Test.DotProductDense(System.ReadOnlySpan`1<Single>, System.ReadOnlySpan`1<Single>, Int32)+0x26e (00007fff`272b3f2e)
mov     rax,qword ptr [rcx]
mov     qword ptr [rsp+30h],rax
xor     ecx,ecx
cmp     dword ptr [rdx+8],0
je      test!Test.DotProductDense(System.ReadOnlySpan`1<Single>, System.ReadOnlySpan`1<Single>, Int32)+0x27e (00007fff`272b3f3e)
mov     rcx,qword ptr [rdx]
mov     qword ptr [rsp+28h],rcx
mov     rdx,rcx
mov     rcx,qword ptr [rsp+30h]
call    Test.DotU(Single*, Single*, Int32) (00007fff`2728a770)
nop
add     rsp,38h
ret
```

### CODE REVIEW

**Comment by jkotas** [comment]
cc @stephentoub

File: src/System.Private.CoreLib/shared/System/ReadOnlySpan.Fast.cs [hunk 1]
```diff
@@ -153,7 +153,13 @@ internal ReadOnlySpan(ref T ptr, int length)
         /// It can be used for pinning and is required to support the use of span within a fixed statement.
         /// </summary>
         [EditorBrowsable(EditorBrowsableState.Never)]
-        public unsafe ref readonly T GetPinnableReference() => ref (_length != 0) ? ref _pointer.Value : ref Unsafe.AsRef<T>(null);
+        public unsafe ref readonly T GetPinnableReference()
+        {
+            // Ensure that the native code has just one forward branch that is predicated-not-taken.
```
Comment (stephentoub) [inline src/System.Private.CoreLib/shared/System/ReadOnlySpan.Fast.cs]: Nit: predicated => predicted?

File: src/System.Private.CoreLib/shared/System/Span.Fast.cs [hunk 2]
```diff
@@ -158,7 +158,13 @@ internal Span(ref T ptr, int length)
         /// It can be used for pinning and is required to support the use of span within a fixed statement.
         /// </summary>
         [EditorBrowsable(EditorBrowsableState.Never)]
-        public unsafe ref T GetPinnableReference() => ref (_length != 0) ? ref _pointer.Value : ref Unsafe.AsRef<T>(null);
+        public unsafe ref T GetPinnableReference()
+        {
+            // Ensure that the native code has just one forward branch that is predicated-not-taken.
```
Comment (stephentoub) [inline src/System.Private.CoreLib/shared/System/Span.Fast.cs]: Same

**Comment by jkotas** [comment]
> Are there code patterns where callers might say "if (span.Length > 0) { pin span here }"? And if so would the redundant length check be properly elided ?

RyuJIT is not able to optimize this well today (https://github.com/dotnet/coreclr/issues/15476 and related issues). This change is not changing that.

**Comment by jkotas** [comment]
2.2 is done at this point. I expect the ship-room is going to question the servicing backport of this fix if I submitted the backport. It will be something like: If there is material impact on Kestrel, rollback the offending change in Kestrel (it is lower risk). If there is no material impact on Kestrel, we are good, no need to backport anything.

**Pull request merged**